### PR TITLE
Defer more configurations

### PR DESCRIPTION
**Pull request by Goooler**

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -1,21 +1,15 @@
...
 pluginManagement {
-  plugins {
-    id "com.gradle.enterprise" version "3.12.4"
-    id "com.gradle.common-custom-user-data-gradle-plugin" version "1.9"
-    id "org.asciidoctor.jvm.convert" version "3.3.2"
-    id "net.nemerosa.versioning" version "3.0.0"
-    id "io.github.gradle-nexus.publish-plugin" version "1.2.0"
-    id "com.github.ben-manes.versions" version "0.46.0"
-    id "biz.aQute.bnd.builder" version "6.4.0"
-    id "io.spring.nohttp" version "0.0.11"
```
[the remark this answers is not in the document]
Comment (Goooler) [reply settings.gradle:12]: I don't see `biz.aQute.bnd.builder` is used in settings.gradle, we can apply necessary plugins in settings.gradle.

Version catalog is the next step to migrate to.

File: settings.gradle [hunk 2]
```diff
@@ -92,7 +86,7 @@ include "spock-guice"
 include "spock-junit4"
 include "spock-testkit"
 
-def variant = System.getProperty("variant") as BigDecimal ?: 2.5
+def variant = providers.systemProperty("variant").getOrElse("2.5").toBigDecimal()
```
[the remark this answers is not in the document]
Comment (Goooler) [reply settings.gradle:89]: The gradleProperty method was introduced in Gradle 6.2 and improves compatibility with Gradle's configuration cache.

File: settings.gradle [hunk 3]
```diff
@@ -1,21 +1,15 @@
...
-    id "com.gradle.common-custom-user-data-gradle-plugin" version "1.9"
-    id "org.asciidoctor.jvm.convert" version "3.3.2"
-    id "net.nemerosa.versioning" version "3.0.0"
-    id "io.github.gradle-nexus.publish-plugin" version "1.2.0"
-    id "com.github.ben-manes.versions" version "0.46.0"
-    id "biz.aQute.bnd.builder" version "6.4.0"
-    id "io.spring.nohttp" version "0.0.11"
+  repositories {
+    gradlePluginPortal()
+    mavenCentral()
```
[the remark this answers is not in the document]
Comment (Goooler) [reply settings.gradle:5]: Reverted in 0d10cd536632d9d13df5c835f7990569c7019e8a.

File: settings.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (Vampire) [reply settings.gradle:12]: > I don't see biz.aQute.bnd.builder is used in settings.gradle, we can apply necessary plugins in settings.gradle.

Of course it is not.
I think you misunderstand `pluginManagement { plugins { ... } }`.
It does not apply anything.
It just defines default versions for plugins if they are applied somewhere else in the build scripts without version.
It is applied in `core.gradle`, with your changes it is now additionally applied to the root project.
Moving the versions out from `pluginManagement { plugins { ... } }` to the usages is a step in the opposite direction of version catalogs.
First there was you "apply the plugin with version", then there was "you can centralize all plugin version declarations using `pluginManagement { plugins { ... } }` and apply them without version", and now the next evolution step are version catalogs.

File: settings.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (Vampire) [reply settings.gradle:89]: Not anymore.
It was necessary in the beginning to use those explicitly so that the configuration cache recognizes them as inputs.
But by now there was code added in Groovy that instruments the whole build script classpaths using bytecode manipulation to also capture all libraries and old plugins and whatever, that reads system properties, environment variables, files, calls external processes and so on.
So for the configuration cache to recognize them, those providers are also not necessary anymore.
Their main usecase now really is, if you need them as `Provider` to feed a `Property`.

File: settings.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (Vampire) [reply settings.gradle:5]: Mind adding it to `build-logic/settings.gradle` too, to be consistent?

[the remark this answers is not in the document]
Comment (Goooler) [reply settings.gradle:5]: 5dbde2b09cd1547d1860315308a29629676d25f1

File: settings.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (Goooler) [reply settings.gradle:12]: I see what you meant. With version catalog declared in `libs.versions.toml`, we can't apply them in `pluginManagement { plugins { ... } }`, right?

File: settings.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (Goooler) [reply settings.gradle]: Revert it for now in 6b94ee1994050af2b0acc7667fa04d04ca2b0611.

File: settings.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (Vampire) [reply settings.gradle:12]: WIth `libs.versions.toml` `pluginManagement { plugins { ... } }` is not necessary anymore.
Again, `pluginManagement { plugins { ... } }` is only defining default versions to be able to apply in the remaining build scripts without version.
With the version catalog, you have the plugins and versions in the version catalog, so `pluginManagement { plugins { ... } }` is then obsolete.

[the remark this answers is not in the document]
Comment (Goooler) [reply settings.gradle]: Alright, just droping them.

**Review by Vampire — APPROVED** [review]
LGTM, thanks

**Comment by Vampire** [comment]
I had a quick look at the build scan.
You greatly improved the task configuration avoidance situation.
But there are still 12 tasks created immediately and 96 registered tasks realized during configuration phase in a typical build run:
https://scans.gradle.com/s/dqgnepfnrzpk2/performance/configuration#summary-task-created-immediately
Do you want to have a look at those too?

**Comment by Goooler** [comment]
Seems the last 8 ones are created in `SigningExtension#sign`, I'm not sure how to defer them.

File: gradle/publishMaven.gradle [hunk 4]
```diff
@@ -7,14 +7,12 @@ ext {
...
-          from(components.java)
-          artifact(sourcesJar)
-          artifact(javadocJar)
-        }
+    register("maven", MavenPublication) {
+      pom.packaging = project.ext.packaging
+      if (project.ext.packaging != "pom") {
+        from(components.java)
+        artifact(sourcesJar)
+        artifact(javadocJar)
```
Comment (Goooler) [inline gradle/publishMaven.gradle:15]: One more thing, the bom module needs `components.javaPlatform` instead of `components.java`, and `java-platform` plugin should be applied in bom modules.

Comment (Goooler) [reply gradle/publishMaven.gradle]: My fault, after applying db00f59d20eddca4b4c3ec62e72f6255f8eea41e & 0ec2232c8c40840d985c436dcb8f83007e3f5fa6, local run:

```sh
./gradlew publishToMavenLocal

tree ~/.m2/repository/org/spockframework/spock-bom/2.4-groovy-2.5-SNAPSHOT/
/Users/snebula/.m2/repository/org/spockframework/spock-bom/2.4-groovy-2.5-SNAPSHOT/
├── maven-metadata-local.xml
├── spock-bom-2.4-groovy-2.5-SNAPSHOT-javadoc.jar
├── spock-bom-2.4-groovy-2.5-SNAPSHOT-sources.jar
├── spock-bom-2.4-groovy-2.5-SNAPSHOT.jar
├── spock-bom-2.4-groovy-2.5-SNAPSHOT.module
└── spock-bom-2.4-groovy-2.5-SNAPSHOT.pom
```

Droping them for now, it looks normal.

```sh
tree ~/.m2/repository/org/spockframework/spock-bom/2.4-groovy-2.5-SNAPSHOT/
/Users/snebula/.m2/repository/org/spockframework/spock-bom/2.4-groovy-2.5-SNAPSHOT/
├── maven-metadata-local.xml
└── spock-bom-2.4-groovy-2.5-SNAPSHOT.pom
```

**Review by leonard84 — DISMISSED** [review]

**Comment by Vampire** [comment]
> Seems the last 8 ones are created in `SigningExtension#sign`, I'm not sure how to defer them.

Unfortunately you cannot: https://github.com/gradle/gradle/issues/18477

@Goooler if you don't want to improve it more, I'm also fine with the current state.
The many tasks that get "Created during configuration" are things like `jar { ... }` to configure the `jar` task instead of `tasks.named('jar') { ... }` or 
```groovy
configureJavadoc(javadoc)
def configureJavadoc(task, boolean root = false) {
  configure(task) {
    ...
  }
}
```
instead of
```groovy
configureJavadoc(tasks.named('javadoc'))
def configureJavadoc(taskProvider, boolean root = false) {
  taskProvider.configure {
    ...
  }
}
```

If you want to stop here, that's fine, if you want to improve it more, that's great.
Basically the "Created during configuration" number should always be 0 except for exceptional edge cases. :-)
Just tell us whether we should merge the current state or whether you want to put some more effort into it.

**Comment by Goooler** [comment]
Good catch!

**Comment by Goooler** [comment]
Refined all I can see.

File: spock-core/core.gradle [hunk 5]
```diff
@@ -80,10 +80,10 @@ processResources {
   filter(org.apache.tools.ant.filters.ReplaceTokens, tokens: tokens)
 }
 
-task coreConsole(type: JavaExec,
-                 description: 'Start a groovy Console with Spock Core Classpath, useful for AST-Inspection') {
+tasks.register("coreConsole", JavaExec) {
+  description = 'Start a groovy Console with Spock Core Classpath, useful for AST-Inspection'
   main = variant == 2.5 ? "groovy.ui.Console" : "groovy.console.ui.Console"
-  classpath = sourceSets.main.runtimeClasspath + configurations.coreConsoleRuntime
+  classpath(sourceSets.named("runtimeClasspath").map {it.runtimeClasspath }, configurations.named("coreConsoleRuntime"))
```
Comment (leonard84) [inline spock-core/core.gradle:86]: Is this really correct shouldn't it be: 
```suggestion
  classpath(sourceSets.named("main").map {it.runtimeClasspath }, configurations.named("coreConsoleRuntime"))
```
Incidentally this feels like it is the reverse of the typesafe accessor change.

Comment (Goooler) [reply spock-core/core.gradle:86]: My typo.

**Review by leonard84 — APPROVED** [review]

File: spock-core/core.gradle (hunk 5 above)
Comment (Vampire) [reply spock-core/core.gradle]: That feels like it because it is Groovy DSL.
Groovy DSL is longer around than Kotlin DSL.
So accessors there are often eager ones while in Kotlin DSL the accessors are lazy.
While actually for `sourceSets` the change would not strictly be necessary.
The only container really benefiting from the new api currently are task containers.
But it also does not hurt too much to always use the new API and be future-proof. :-)
And if someone invests the time to rewrite the build scripts in Kotlin DSL, it will probably also look more accessor-y again. :-)

File: build.gradle [hunk 6]
```diff
@@ -346,25 +350,25 @@ task tagRelease(type: Exec) {
...
+tasks.register("groovydoc", Groovydoc) {
   docTitle "Spock Framework API Documentation ($variantLessVersion)"
   windowTitle "Spock Framework API Documentation ($variantLessVersion)"
   destinationDir file("build/groovydoc/$variantLessVersion")
-  source subprojects.groovydoc.source
-  classpath = files(subprojects.groovydoc.classpath)
-  groovyClasspath = projects.spockCore.dependencyProject.groovydoc.groovyClasspath
+  source subprojects.collect { project -> project.tasks.named("groovydoc").map { it.source } }
+  classpath = files(subprojects.collect { project -> project.tasks.named("groovydoc").map { it.classpath } })
+  groovyClasspath = projects.spockCore.tasks.named("groovydoc").map { it.groovyClasspath }
```
Comment (Vampire) [inline build.gradle:368]: This does not work

**Review by Vampire — CHANGES_REQUESTED** [review]
There are 5 more that can be avoided to be realized eagerly:
```patch
diff --git a/build.gradle b/build.gradle
index e3a76275a9..f80a64634c 100755
--- a/build.gradle
+++ b/build.gradle
@@ -385,7 +385,7 @@ destinationDir file("build/groovydoc/$variantLessVersion")
   }
 }

-asciidoctor {
+tasks.named("asciidoctor") {
   configurations 'asciidoctorExtensions'
   sourceDir = "docs"
   baseDirFollowsSourceDir()
diff --git a/gradle/ide.gradle b/gradle/ide.gradle
index 8406d80122..b0f3ca81d8 100644
--- a/gradle/ide.gradle
+++ b/gradle/ide.gradle
@@ -41,6 +41,6 @@
   }
 }

-ideaWorkspace {
+tasks.named("ideaWorkspace") {
   outputFile = file(outputFile.path.replace(".iws", "-${variant}.iws"))
 }
diff --git a/spock-core/core.gradle b/spock-core/core.gradle
index e9b32791a3..ea620d526a 100755
--- a/spock-core/core.gradle
+++ b/spock-core/core.gradle
@@ -74,7 +74,7 @@ api platform(libs.junitBom)
   }
 }

-processResources {
+tasks.named("processResources") {
   def tokens = [version: version.toString(), minGroovyVersion: minGroovyVersion, maxGroovyVersion: maxGroovyVersion]
   inputs.property "tokens", tokens
   filter(org.apache.tools.ant.filters.ReplaceTokens, tokens: tokens)
diff --git a/spock-specs/mock-integration/mock-integration.gradle b/spock-specs/mock-integration/mock-integration.gradle
index f7687af9f1..fec15942a0 100755
--- a/spock-specs/mock-integration/mock-integration.gradle
+++ b/spock-specs/mock-integration/mock-integration.gradle
@@ -36,4 +36,4 @@
   }
 }

-check.dependsOn(tasks.withType(Test))
+tasks.named("check") { dependsOn(tasks.withType(Test)) }
diff --git a/spock-spring/spring.gradle b/spock-spring/spring.gradle
index f68399edd6..cf8a058f90 100755
--- a/spock-spring/spring.gradle
+++ b/spock-spring/spring.gradle
@@ -88,7 +88,7 @@ implementation project(project.path)
           }
         }
       }
-      check.dependsOn testTaskName
+      tasks.named("check") { dependsOn testTaskName }
     }
   }
 }
```

File: build.gradle (hunk 6 above)
Comment (Goooler) [reply build.gradle]: Fixed in 15998094c8041336c765c53fd0f1eef6f44dad10.

Comment (Vampire) [reply build.gradle]: No, still does not work

Comment (Goooler) [reply build.gradle]: c98ccca82d4fdb76b2cf7bc37eea016f3e3ef842

Comment (Goooler) [reply build.gradle]: It broken on master branch:

```
* What went wrong:
Some problems were found with the configuration of task ':groovydoc' (type 'Groovydoc').
  - Type 'org.gradle.api.tasks.javadoc.Groovydoc' property 'access' doesn't have a configured value.
    
    Reason: This property isn't marked as optional and no value has been configured.
    
    Possible solutions:
      1. Assign a value to 'access'.
      2. Mark property 'access' as optional.
    
    Please refer to https://docs.gradle.org/8.0.2/userguide/validation_problems.html#value_not_set for more details about this problem.
```

Comment (Vampire) [reply build.gradle]: Well, it still fails, but at least now it fails with the same error as `upstream/master` :-D

**Review by Vampire — DISMISSED** [review]

**Comment by Goooler** [comment]
Thanks for your reviews.

**Comment by Vampire** [comment]
Thanks for your work :-)

**Review by Vampire — APPROVED** [review]
